### PR TITLE
Split history replication common logic into mapper file

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Split history replication common logic into mapper file

**Why?**
Better code reusability

**How did you test it?**
existing tests

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

**Comment by wxing1292** [comment]
review #4807 first

File: service/history/ndc/mutable_state_mapper.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	newMutableState, err := stateBuilder.ApplyEvents(
+		ctx,
+		task.getNamespaceID(),
+		uuid.New().String(),
+		*task.getExecution(),
+		task.getEvents(),
+		task.getNewEvents(),
+	)
+	if err != nil {
+		task.getLogger().Error(
```
Comment (xwduan) [inline service/history/ndc/mutable_state_mapper.go:175]: Nit: Seems the caller of this method is also logging the error. Will this produce duplicated error log?

File: service/history/ndc/history_replicator.go [hunk 2]
```diff
@@ -732,24 +645,20 @@ func (r *HistoryReplicatorImpl) applyNonStartEventsResetWorkflow(
...
+	mutableState, newMutableState, err := r.mutableStateMapper.ApplyEvents(ctx, wfContext, mutableState, task)
 	if err != nil {
 		task.getLogger().Error(
-			"nDCHistoryReplicator unable to apply events when applyNonStartEventsResetWorkflow",
+			"HistoryReplicator::applyNonStartEventsResetWorkflow unable to apply events",
 			tag.Error(err),
 		)
 		return err
 	}
+	if newMutableState != nil {
```
Comment (xwduan) [inline service/history/ndc/history_replicator.go:656]: Q: So we ignored this case before this change? Could be a bug?

**Review by xwduan — APPROVED** [review]

File: service/history/ndc/mutable_state_mapper.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+func (m *MutableStateMapperImpl) FlushBufferEvents(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	task replicationTask,
+) (workflow.MutableState, struct{}, error) {
+	flusher := m.newBufferEventFlusher(wfContext, mutableState, task.getLogger())
+	_, mutableState, err := flusher.flush(ctx)
```
Comment (yux0) [inline service/history/ndc/mutable_state_mapper.go:93]: do you need to return the context?

File: service/history/ndc/history_replicator.go (hunk 2 above)
Comment (wxing1292) [reply service/history/ndc/history_replicator.go:656]: before we impl replication task batching, this case will not happen.

File: service/history/ndc/mutable_state_mapper.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+		newStateBuilder:       newStateBuilder,
+	}
+}
+
+func (m *MutableStateMapperImpl) FlushBufferEvents(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	task replicationTask,
+) (workflow.MutableState, struct{}, error) {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/ndc/mutable_state_mapper.go:91]: for function signature consistency, at least for now. i am still experimenting whether <- is useful or not

File: service/history/ndc/mutable_state_mapper.go (hunk 3 above)
Comment (wxing1292) [reply service/history/ndc/mutable_state_mapper.go:93]: context is not expected to change, but mutable state can, e.g. ms rebuild due to conflict

File: service/history/ndc/mutable_state_mapper.go (hunk 1 above)
Comment (wxing1292) [reply service/history/ndc/mutable_state_mapper.go:175]: good point, let me update

**Review by yux0 — APPROVED** [review]

**Pull request merged**